### PR TITLE
Full Join Logic for Dimension Mappings

### PR DESCRIPTION
**Pull request by daniel-thom**
Completes JIRA Story DSGRID-186

This PR also fixes handling of dimension associations. The associations registered to existing projects did not work because the data files were not copied to the registry. We can patch this by manually uploading the files to S3.

### CODE REVIEW

File: dsgrid/config/project_config.py
Comment (daniel-thom) [inline dsgrid/config/project_config.py:60]: Many of our models unnecessarily define types with `Optional`. If a default value is provided then we don't want them declared `Optional`. That type means that it can be `None` or the real type.

File: dsgrid/config/project_config.py
Comment (daniel-thom) [inline dsgrid/config/project_config.py:168]: @mooneyme Is this field now defunct?

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [inline dsgrid/registry/project_registry_manager.py:285]: @mooneyme Please confirm the logic here.

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [inline dsgrid/registry/project_registry_manager.py:308]: If someone has a better suggestion, please raise it.

File: dsgrid/config/project_config.py
Comment (elainethale) [inline dsgrid/config/project_config.py:343]: Are we calling these association mappings now (instead of base-to-base mappings)? (Maybe related to @daniel-thom 's question to @mooneyme about a possibly defunct-base-to-base field?)

File: dsgrid/config/project_config.py
Comment (elainethale) [inline dsgrid/config/project_config.py:386]: Flagging another instance of base-to-base

Comment (elainethale) [reply dsgrid/config/project_config.py:386]: Looking farther down, we do of course have dataset-base to project-base mappings. Then I think we're calling within project base-to-base mappings "association mappings" instead. Is that correct? 

Without looking a lot closer I can't tell which kind of mapping (of the two I just listed) is meant here.

File: dsgrid/registry/project_registry_manager.py
Comment (elainethale) [inline dsgrid/registry/project_registry_manager.py:374]: This tells me we're dropping values in df not found in the mapping.

File: dsgrid/registry/project_registry_manager.py
Comment (elainethale) [inline dsgrid/registry/project_registry_manager.py:377]: And here we're just switching to using to_id as our identifier. It could be that the mapping represents an aggregation; I agree that for this use case we don't care/don't want to actually deal with duplicates, but if/when this moves elsewhere we'll probably want to encode somewhere (the mapping_config?) whether we want to perform a sum or some other aggregation to collapse duplicates when we are done.

**Review by elainethale — APPROVED** [review]
Looks good to me. I have some clarification questions about base-to-base mappings. Otherwise I am super happy to see these features go in.

Note that I have not tried to run the tests myself nor did I go too deep into the queries/logic that is doing the actual checking (although I did spend some time reading them to get a basic sense of the intention and methods).

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:343]: Here is my interpretation of things:
- A `dimension mapping` maps a `from_id` to a `to_id`. Example: dataset calls a dimension record something slightly different than the project.
- A `dimension association` in the project.toml specifies what combinations of records are required to be present in the load data table. This association can have two or more columns. If the dataset does have a given combination then it must still have a row for the combination, but the ID is null.

I saw that @mooneyme deleted base-to-base mappings in the project files and am assuming that those are now obsolete.

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:374]: My addition in https://github.com/dsgrid/dsgrid/pull/101 requires that a dimension mapping include all records, so this code is currently correct. However, I should have included an assert to make sure the length didn't change.

I actually think my earlier change is incorrect. We don't need to require that all records are covered in the mapping.

File: dsgrid/registry/project_registry_manager.py
Comment (lixiliu) [reply dsgrid/registry/project_registry_manager.py:308]: If lookup is a subset of records, some flavor of this might work:
`records[[type1.value, type2.value]].except(lookup[[type1.value, type2.value]])`

File: dsgrid/registry/project_registry_manager.py
Comment (lixiliu) [inline dsgrid/registry/project_registry_manager.py:294]: Tip: if `pdim1_ids` and `pdim12_ids` happen to be a col from other dataframes, you can do `pdim1_ids.crossJoin(pdim2_ids)` to get the product dataframe.

File: dsgrid/config/project_config.py
Comment (lixiliu) [reply dsgrid/config/project_config.py:343]: A project is a combination of datasets minus data it does not want. If we have to generate a `dimension association` from datasets anyway, then I think it's easier and safer for project to specify what it does not want from each dataset and have the association be generated based on that.

Comment (mooneyme) [reply dsgrid/config/project_config.py:343]: That is right Dan.

File: dsgrid/config/project_config.py
Comment (mooneyme) [reply dsgrid/config/project_config.py:168]: this is now what we are calling an association

File: dsgrid/registry/project_registry_manager.py
Comment (mooneyme) [reply dsgrid/registry/project_registry_manager.py:377]: Yes agreed Elaine.

Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:377]: @elainethale I don't think this comment is relevant to this PR (or at least there are no changes associated with it), so I'm resolving it. I'll assume that you or Meghan will make appropriate plans for any changes to mapping configs.

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:386]: All references to base-to-base mappings will be deleted in the next push to this PR.

File: tests/test_dataset_registration.py
Comment (mooneyme) [inline tests/test_dataset_registration.py:22]: Do these tests belong in the test_dataset_registration.py? Because really we need to test this at dataset submission to project. Right?

Comment (daniel-thom) [reply tests/test_dataset_registration.py:22]: It is running the submit_dataset command. I'll rename the file.

File: tests/data/dimension_associations/subsector__metric.csv
Comment (mooneyme) [inline tests/data/dimension_associations/subsector__metric.csv:3]: Why do you have associations in dsgrid and not dsgrid-test-data?

Comment (daniel-thom) [reply tests/data/dimension_associations/subsector__metric.csv:3]: dsgrid-test-data has toy versions of projects and dataset that comprise real registries. This is just a few independent files to test the DimensionAssociation class by itself.

File: dsgrid/config/dataset_schema_handler_base.py
Comment (lixiliu) [inline dsgrid/config/dataset_schema_handler_base.py:234]: I would add how="left". Default is an inner join (https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.join.html), if `from_id` is partial of a col, df.count() will shrink.

File: dsgrid/config/dataset_schema_handler_one_table.py
Comment (lixiliu) [inline dsgrid/config/dataset_schema_handler_one_table.py:94]: Since there's no collapsing/aggregation in _remap_dimension_columns, this will contain duplicates if to_id in maps contain duplicates, which seem to go against what the name of the func implies

File: dsgrid/registry/project_registry_manager.py
Comment (lixiliu) [inline dsgrid/registry/project_registry_manager.py:342]: We want symmetric_difference here right since we are checking that the mapped dimensions match exactly what's expected at project?

Comment (lixiliu) [reply dsgrid/registry/project_registry_manager.py:342]: If not symmetric_difference, I think it's `diff = p_dim_ids.difference(d_dim_ids)`

File: dsgrid/config/dataset_schema_handler_base.py
Comment (daniel-thom) [reply dsgrid/config/dataset_schema_handler_base.py:111]: If `to_id` is None then the record cannot be present in the pivoted columns. The dict returned from this method will only include actual columns. This is just for the check on line 114. Let me know if I'm missing something.

I do think there is a bug here. `to_id` may be `""` instead of `None` because the `AssociationTableRecordModel` defines the field as a string. I think we need to coerce these values to `None`.

File: dsgrid/config/dataset_schema_handler_base.py
Comment (daniel-thom) [reply dsgrid/config/dataset_schema_handler_base.py:234]: In one of our meetings last week we discussed this and decided that we do want an inner join. The mapping should include all records. Note that line 239 guards against this point.

File: dsgrid/config/dataset_schema_handler_one_table.py
Comment (daniel-thom) [reply dsgrid/config/dataset_schema_handler_one_table.py:94]: Yeah, this is a bug. It should be sufficient to append a `.distinct()` to the end of this, right?

File: dsgrid/config/dataset_schema_handler_base.py
Comment (lixiliu) [reply dsgrid/config/dataset_schema_handler_base.py:111]: I meant to say that `diff = columns.symmetric_difference(mapping)` should be done before anything is removed from the mapping. Also, shouldn't columns be compared against the `from_id` only in the mapping?

File: dsgrid/config/dataset_schema_handler_base.py
Comment (lixiliu) [reply dsgrid/config/dataset_schema_handler_base.py:234]: Right I saw line 239. If records are already checked to ensure that all from_id exists, then that makes sense.

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:342]: I'm not sure how I messed this one up. The exact project-required records are known here. I'll fix.

File: dsgrid/config/project_config.py
Comment (elainethale) [inline dsgrid/config/project_config.py:336]: Everywhere we have dimension: DimensionType, shouldn't the name be dimension_type?

File: dsgrid/config/project_config.py
Comment (elainethale) [inline dsgrid/config/project_config.py:360]: _has_mapping(from_dimension_type, to_dimension_type, mappings):

?

File: dsgrid/registry/project_registry_manager.py
Comment (elainethale) [inline dsgrid/registry/project_registry_manager.py:302]: Will this grab the right associations if we have something like:

| data_source | sector |
|------------- | ------ |
| resstock | res |

| sector | subsector |
| ------ | ---------- |
| res      | SingleFamily |
| res     | MultiFamiliy |

?

File: dsgrid/utils/files.py
Comment (elainethale) [inline dsgrid/utils/files.py:75]: Seems like this needs to be removed before merging?

File: tests/data/dimension_associations/model_year__metric.csv
Comment (elainethale) [inline tests/data/dimension_associations/model_year__metric.csv:11]: Why do we need associations for model_years?

File: tests/data/dimension_associations/scenario__subsector.csv
Comment (elainethale) [inline tests/data/dimension_associations/scenario__subsector.csv:14]: This seems redundant with the scenario,data_source and data_source,subsector associations. Is that what's intended?

File: dsgrid/config/dataset_schema_handler_one_table.py
Comment (daniel-thom) [reply dsgrid/config/dataset_schema_handler_one_table.py:35]: 1. I think there may have been more than one function call at some point in time.
2. You get to choose the name that shows up in the logs with `Timer`. I may have preferred no underscore. 

I'm open to changing these now.

File: dsgrid/config/dimension_associations.py
Comment (daniel-thom) [reply dsgrid/config/dimension_associations.py:66]: This file does not use `dimension_type`. I used `dimension` here for brevity because there shouldn't be any confusion with `dimension_type` vs `dimension_config`, which could exist in other files. I still prefer the shorter name but am open to changing it.

File: dsgrid/config/dimension_associations.py
Comment (daniel-thom) [reply dsgrid/config/dimension_associations.py:83]: All keys are already tuples. The point of this code block is the case where the files have 3 dimensions but the caller asks for 2 and those 2 dimensions exist in 2 different files. I'm a bit concern that there is some corner-case problem not covered.

File: dsgrid/config/dimension_associations.py
Comment (daniel-thom) [reply dsgrid/config/dimension_associations.py:131]: Hmm. Would it be an error if any records are dropped by an inner join? I could be missing something but it seems like that shouldn't happen?

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:336]: I'm not sure I agree with the statement globally, but I do agree here. If the line length limit forced this to be
```
def get_base_to_supplemental_dimension_mappings_by_types(
    self,
    dimension: DimensionType
):
```
then I would want to use the shorter name.

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:302]: Does this answer your question? Also, you can directly load the dimension associations for StandardScenarios without project or datasets if you'd like to experiment.
```
(Pdb) self.get_associations(DimensionType.DATA_SOURCE, DimensionType.SECTOR).show()
+-----------+------+
|data_source|sector|
+-----------+------+
|   comstock|   com|
|   comstock|   res|
+-----------+------+
(Pdb) self.get_associations(DimensionType.SECTOR, DimensionType.SUBSECTOR).show()
+------+--------------------+
|sector|           subsector|
+------+--------------------+
|   com|FullServiceRestau...|
|   com|            Hospital|
|   com|          LargeHotel|
|   com|         LargeOffice|
|   com|        MediumOffice|
|   com|          Outpatient|
|   com|       PrimarySchool|
|   com|QuickServiceResta...|
|   com|          SmallHotel|
|   com|         SmallOffice|
|   com|    StandaloneRetail|
|   com|           StripMall|
|   com|           Warehouse|
|   res|    MidriseApartment|
+------+--------------------+
(Pdb) self.get_associations_by_data_source("comstock", DimensionType.SECTOR, DimensionType.SUBSECTOR).show()
+--------------------+------+
|           subsector|sector|
+--------------------+------+
|FullServiceRestau...|   com|
|            Hospital|   com|
|          LargeHotel|   com|
|         LargeOffice|   com|
|        MediumOffice|   com|
|          Outpatient|   com|
|       PrimarySchool|   com|
|QuickServiceResta...|   com|
|          SmallHotel|   com|
|         SmallOffice|   com|
|    StandaloneRetail|   com|
|           StripMall|   com|
|           Warehouse|   com|
|    MidriseApartment|   res|
+--------------------+------+
```

File: dsgrid/utils/files.py
Comment (daniel-thom) [reply dsgrid/utils/files.py:75]: thanks

File: tests/data/dimension_associations/model_year__metric.csv
Comment (daniel-thom) [reply tests/data/dimension_associations/model_year__metric.csv:11]: I copied everything from StandardScenarios for test purposes - not necessarily because it makes sense for this project. Maybe I should remove some of the associations.

File: tests/test_dimension_associations.py
Comment (mooneyme) [inline tests/test_dimension_associations.py:20]: @daniel-thom what if I want to get all the associations for a specific data_source? Where are we doing the join?
`da.get_associations_by_data_source("comstock").show()` is an empty table.

File: tests/test_dimension_associations.py
Comment (mooneyme) [inline tests/test_dimension_associations.py:15]: Even though we don't provide an association table for this, I still want to be able to see which Sectors are in each Geography (or Time). Remember that if it isn't provided, then we assume a full join between the combos of dimensions.

File: tests/test_dimension_associations.py
Comment (mooneyme) [inline tests/test_dimension_associations.py:1]: In this test script, we need to add some tests for bad associations.

File: tests/test_register_project.py
Comment (mooneyme) [inline tests/test_register_project.py:47]: Does it make sense that this is here? the return should be in dsgrid code, not in tests, right?

File: dsgrid/config/dimension_associations.py
Comment (mooneyme) [inline dsgrid/config/dimension_associations.py:58]: Shouldn't we be validating the associations before we return?

Comment (mooneyme) [reply dsgrid/config/dimension_associations.py:58]: Also, where is the full join logic happening?

File: dsgrid/config/dimension_associations.py
Comment (daniel-thom) [inline dsgrid/config/dimension_associations.py:176]: @elainethale @mooneyme @lixiliu I think that producing a single table of dimension associations is something like this. I'm unsure about `how="cross"` vs `how="outer"`. Results appear the same in limited testing.

I think that we could implement all checks by keeping this separate from the full combinations of project records. We could still iterate over pairs of combinations in project_registry_manager and check if they exist here. If there was somehow implicit relationships that still need to be accounted for, that wouldn't work.

Let me know what you think. If this seems right to you then I can implement the logic today. Thanks.

Comment (mooneyme) [reply dsgrid/config/dimension_associations.py:176]: @daniel-thom here is what I have been playing around with. Given our current test, this works as we would expect. However I think there are a few case study tests we need to test out to make sure it behaves as expected under every potential scenario.

```
    def get_full_join_by_data_source(self, project_config, data_source):
        tables = []
        for dims, val in self._associations.items():
            if DimensionType.DATA_SOURCE in dims:
                continue
            table = self.get_associations_by_data_source(data_source, *dims)
            tables.append(table)

        table = tables[0]
        if len(tables) > 1:
            for i, other in enumerate(tables[1:]):
                print(i)
                on_columns = list(set(other.columns).intersection(table.columns)) # TODO: is there potentially an issue here with the ordering of the tables to lead to a "non-overlap" on_columns?
                if on_columns:
                    table_columns = table.columns
                    table_distinct_cols_count =  table.select(table.columns).distinct().count() 
                    table = table.join(other, on=on_columns, how="outer")
                    assert table.select(other.columns).distinct().count() == other.select(other.columns).distinct().count() 
                    assert table.select(table_columns).distinct().count() == table_distinct_cols_count 
                else:
                    # TODO: need to test this with the right combinations of supplied associations
                    table = table.crossJoin(other)
                    
        # TODO: in our current test using standard scenarios dimensions, we only have 5 dimensions in the columns, but 8 tables are joined together. This has to do with the full circle chain of associations provided. We may be able to save time and skip these as long as we do some sanity checks?? Unclear ATM.
        
        all_dimensions = set(d.value for d in DimensionType)
        missing_dimensions = all_dimensions.difference(table.columns)
        for dim in missing_dimensions:
            if dim not in ("data_source", "time"): # TODO: how do we want to add in time dimension if we don't have records for them?
                table_count = table.count()
                # TODO: @dtom is there a better/preferred way to get the records of a dimension?
                other = project_config.get_base_dimension(DimensionType(dim)).get_records_dataframe().select("id as geography")
                table = other.crossJoin(table.withColumnRenamed("id", dim))
                assert table.count() == other.count() * table.count()
        
        table = table.withColumn("data_source", lit(data_source))
        return table
```

Comment (mooneyme) [reply dsgrid/config/dimension_associations.py:176]: Also just a note. I found that it is an "outer join" when combining association tables provided that have an overlapping dimension column, however it is a "cross join" when we are joining association tables without an overlapping column or dimensions that were not provided with an association table (e.g., geography).

File: dsgrid/config/dimension_associations.py
Comment (daniel-thom) [inline dsgrid/config/dimension_associations.py:27]: @mooneyme I refactored this class to store a single table containing all associations. Please check the joins again. I left your asserts out. I'm uncertain if they are needed. I typically use asserts to guard against coding errors and not bad inputs. If you were guarding against bad input then they need to be added back.

I did not include a join of the project config records because I think it's better to keep the code modular. You can still create that table very easily in `project_registry_manager.py` in the `_check_dataset_base_to_project_base_mappings` method. The main advantage of this approach is that dimension_associations.py can be tested without a project.

You will also see that I left the code that performs project/dataset checks by iterating over pairs of dimensions. As discussed, it might be possible to make this faster by comparing one overall table with the dataset table (`dim_table` in the method above).

File: dsgrid/dataset/dataset_schema_handler_base.py
Comment (lixiliu) [inline dsgrid/dataset/dataset_schema_handler_base.py:225]: We need to change how dimension columns get remapped to handle duplicates in from_id.

**Review by lixiliu — CHANGES_REQUESTED** [review]

File: dsgrid/dataset/dataset_schema_handler_base.py
Comment (lixiliu) [reply dsgrid/dataset/dataset_schema_handler_base.py:225]: To be addressed in #108

**Comment by lixiliu** [comment]
@daniel-thom - #108 contains some revision intended for this PR. You should review it (to be merged into this PR) first before making changes.

**Pull request merged**
